Export controller skeleton

elevran opened the pull request:
Actual handlers stubbed (see todo notes for intended implementation method).

References #2 

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
create skeleton for hub ServiceExport controller

Actual handlers stubbed.

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
create skeleton for hub ServiceExport controller

Actual handlers stubbed.

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

davidohana commented on cmd/hub/mcscontroller/main.go:
```diff
@@ -0,0 +1,135 @@
+	var probeAddr string
+	var namespace string
+
+	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
```
not sure the default address should be 8080, in the agent its 8082 hard coded.

davidohana commented on cmd/hub/mcscontroller/main.go:
```diff
@@ -0,0 +1,135 @@
+	opts.BindFlags(flag.CommandLine)
+
+	flag.Parse()
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
```
We should make sure we use the same logging infra in agent and hub. In the agent I used the kzerolog implementation for the logr interface. lets discuss this.

elevran replied:
Agree we should be consistent. I had used the default kubebuilder. We can do this in a separate PR

davidohana replied:
agree, I can do it myself after this PR.

davidohana commented on cmd/hub/mcscontroller/main.go:
```diff
@@ -0,0 +1,135 @@
+	flag.Parse()
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	namespace, err := getWatchedNamespace()
```
what about the command line arg? it should take precedence ..

davidohana commented on cmd/hub/mcscontroller/main.go:
```diff
@@ -0,0 +1,135 @@
+
+	namespace, err := getWatchedNamespace()
+	if err != nil {
+		setupLog.Error(err, "failed to get watched namespace, running in cluster scope")
```
do you want to support this mode? or just early break ..

elevran replied:
Probably not for this iteration, though the rest of the code is namespace scoped by default.

davidohana commented on cmd/hub/mcscontroller/main.go:
```diff
@@ -0,0 +1,135 @@
+		setupLog.Error(err, "failed to get watched namespace, running in cluster scope")
+	}
+
+	setupLog.Info("starting hub MCS Controller", "namespace", namespace)
```
in submariner they usually write logs sentences uppercased. I also prefer the lowercase form but we should be consistent.,

elevran replied:
Changed throughout to capitalize Info and Error messages (but *not* error strings)

davidohana commented on cmd/hub/mcscontroller/main.go:
```diff
@@ -0,0 +1,135 @@
+		Port:                   9443,
+		HealthProbeBindAddress: probeAddr,
+		LeaderElection:         enableLeaderElection,
+		LeaderElectionID:       "386401.lighthouse.submariner.io",
```
should it be const?

elevran replied:
it is autogenerated by kubebuilder. I wouldn't make it a const, though we could.

davidohana commented on cmd/hub/mcscontroller/main.go:
```diff
@@ -0,0 +1,135 @@
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                 scheme,
+		MetricsBindAddress:     metricsAddr,
+		Port:                   9443,
```
also, a const?

elevran replied:
yes and no... It only appears in one place at this point so less of a consistency issue

davidohana commented on cmd/hub/mcscontroller/main.go:
```diff
@@ -0,0 +1,135 @@
+	//+kubebuilder:scaffold:scheme
+}
+
+func main() {
```
the agent also has an uninstall option, I think you will need to support it as well.

elevran replied:
Not sure. Wouldn't kubcetl delete handle it for us? We are not creating additional cluster state that needs to be "installed" and "uninstalled"

davidohana replied:
We create objects and we want to delete them when uninstalling. 
subctl has an "uninstall" option that executes a custom uninstallation procedure in all of its components. 
If you look at the agent's main.go you see that it supports it, so I think  the hub's main shall too.

davidohana replied:
It may be added in a separate PR, just we may want to track it in an issue

elevran replied:
https://github.com/axon-net/lighthouse/issues/19

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,127 @@
+		return r.handleDelete(ctx, &se)
+	}
+
+	return r.handleUpdate(ctx, &se)
```
better call it `handleCreateOrUpdate()`

elevran replied:
changed to just calling it `reconcile`

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,127 @@
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceimports,verbs=create;get;list;update;patch
+
+// Reconcile handles an update to a ServiceExport
+// By design, the event that triggered reconciliation is not passed to the reconciler
```
but it somewhat does .. ?  you resolve the event from the NamespacedName and react to the specific event type in the `handle*()` funcs .. I dont think we can call define this approach as fully state driven

elevran replied:
Added note about needing to be aware of deletion events in a description of the reconciliation logic.
the issue is that client.List won't return the deleted object so we need to explicitly delay deletion (with a finalizer) so we can retrieve the "real" name and namespace without relying on a specific name format (such as name-ns-cluster).
Changed so we do extra work on cleaning up only when DeletionTimeStamp is set, but otherwise call `reconcile` just the same for all operations.

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,127 @@
+	}
+
+	if se.GetDeletionTimestamp() != nil { // service export is marked to be deleted
+		return r.handleDelete(ctx, &se)
```
what happens if deletion or update was made while the  controller is down?

elevran replied:
On start (and restart), the framework send a create event event for every existing object.
ServiceImports of deleted exports will be automatically garbage collected by virtue of having an OwnerRefefence

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,127 @@
+	log := r.Log.WithValues("serviceexport", types.NamespacedName{se.Namespace, se.Name})
+
+	if !controllerutil.ContainsFinalizer(se, serviceExportFinalizer) {
+		controllerutil.AddFinalizer(se, serviceExportFinalizer)
```
I didnt understand the purpose of using a finalizer. Perhaps you can explain it as a code comment?

elevran replied:
Added.

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,127 @@
+// By design, the event that triggered reconciliation is not passed to the reconciler
+// to ensure actions are based on state alone. This approach is referred to as level-based,
+// as opposed to edge-based.
+func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
do we want/need a full reconciliation on startup of the controller? to adapt to state changes that were made offline?
do you see usage for a periodic full reconciliation just to make sure we didnt miss some event?

elevran replied:
switched to more state based reconciliation. Controller is sync with create events on start to pick up existing state.

davidohana reviewed (commented):
completed the first review iteration

elevran committed:
reconcilation loop logic simplified

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
reconcilation loop logic simplified

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
utility functions for lighthouse labels (#18)

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
Merge branch 'export_controller' of github.com:elevran/lighthouse into export_controller

elevran committed:
updated dependencies

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
address review comments

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,169 @@
+		if apierrors.IsNotFound(err) {
+			log.Info("No ServiceExport found")
+		} else {
+			log.Error(err, "Error fetching ServiceExport")
```
Will this cause a retry? Can be some temp error that is resolved later ..

elevran replied:
possibly. 
Currently not retries. Can easily add a `return ctrl.Result{ Requeue: true }` though I'm not sure which of the errors in https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors ought be retriable?

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,169 @@
+	se := mcsv1a1.ServiceExport{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &se); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("No ServiceExport found")
```
In what situation this can happen? Isn't it considered an error as well?

davidohana replied:
Will this happen if for some reason we could not handle the deletion event in time while finalizer was active? What happens then? no recon will occur right?

elevran replied:
> In what situation this can happen? Isn't it considered an error as well?

when the object was fully deleted between enqueuing the notification and calling Get.
Unclear what you're asking in the other comment.

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,169 @@
+)
+
+const (
+	serviceExportFinalizer = "lighthouse.submariner.io/service-export-finalizer"
```
would call it `serviceExportFinalizerName`

davidohana commented on pkg/mcs/lighthouse.go:
```diff
@@ -0,0 +1,91 @@
+// conventions. The type is not strictly needed (hence the lack of receiver
+// object in all calls) but can be used a a scope for shorter function
+// names.
+type Lighthouse struct{}
```
I would call it `LighthouseUtils` at least
Acutally I didn't understand why its better that a package named `lhutils` with static functions.

elevran replied:
It's not better than a `lhutils` package other than avoid proliferation of packages (I wanted to add it to `mcs` so our affected code base is limited).
I could have named the functions with a `Lighthouse` prefix as well. Having a struct just means you can use a shorter name for it. A package would achieve the same (using a shorter import name). 
I can remove this struct altogether if not consistent or style us unexpected.

elevran replied:
moved to new package `lhutil`

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,169 @@
+// owned ServiceImport object will be deleted.
+//
+// Note: ownership will need to be reconsidered if merging imports
+func (r *ServiceExportReconciler) handleDelete(ctx context.Context, se *mcsv1a1.ServiceExport) (ctrl.Result, error) {
```
why do we even need to bother with explicitly removing the finalizer? I recall you said that k8s removes objects with finalizers after some grace period.

elevran replied:
I was wrong. k8s will not delete the object until finalizers is empty

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,169 @@
+		log := r.Log.WithValues("service", lh.GetOriginalObjectName(se.ObjectMeta)).
+			WithValues("cluster", lh.GetOriginalObjectCluster(se.ObjectMeta))
+
+		log.Info("Removing service export")
```
`"Removing service export finalizer"`

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,169 @@
+// - ServiceImports are owned by the corresponding ServiceExport and are thus
+//	 automatically deleted by k8s when their ServiceExport is deleted.
+//
+func (r *ServiceExportReconciler) reconcile(ctx context.Context, se *mcsv1a1.ServiceExport) (ctrl.Result, error) {
```
I think its better to accept a name and namespace of the service explicitly and not the export here.

elevran replied:
wouldn't the first thing to be done in `reconcile` is be to retrieve the ServiceExport? What do we get from passing in the NamespacedName explicitly instead of the SE we just retrieved?
😕

davidohana replied:
another option is to move the finalizer check outside to the calling function

elevran replied:
Changed to pass in the original service and service name (cluster not currently consulted - could be added if needed).

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,169 @@
+//	 automatically deleted by k8s when their ServiceExport is deleted.
+//
+func (r *ServiceExportReconciler) reconcile(ctx context.Context, se *mcsv1a1.ServiceExport) (ctrl.Result, error) {
+	lh := mcs.Lighthouse{}
```
better to create this one time at a static scope ?

elevran replied:
agree. But need to decide if we need it at all or just prefix the functions with `Lighthouse` (I would prefer not to add yet another package just for having lighthouse utilities. Another option is to have these as private functions in the controller.
WDYT?

elevran replied:
moved to a new package

davidohana commented on cmd/hub/mcscontroller/controller.go:
```diff
@@ -0,0 +1,169 @@
+	// @todo: add the NamespacedName of the ServiceExport to be compatible with Reconcile?
+	log := r.Log.WithValues("service", name).WithValues("cluster", lh.GetOriginalObjectCluster(se.ObjectMeta))
+
+	if !controllerutil.ContainsFinalizer(se, serviceExportFinalizer) && se.GetDeletionTimestamp() == nil {
```
I think this should be done in the loop iteration all exports and not only for the specific export that triggered the recon.

elevran replied:
won't we get a notification for them any way and then handle it? Why do it ahead of time? We haven't done any work on a SE before it has a finalizer so if it gets deleted before we have a chance - no harm done. 
I think we can live with that race condition.

elevran replied:
Moved to Reconcile but still done per object and not trying to process all relevant at once.

davidohana commented on pkg/mcs/lighthouse.go:
```diff
@@ -0,0 +1,91 @@
+// Label the object with the expected source object identity labels.
+// Note: metav1.SetMetaDataLabel added in v0.20.0
+func (*Lighthouse) Label(objmd *metav1.ObjectMeta, name, ns, cluster string) {
+	objmd.GetLabels()[lhconst.LighthouseLabelSourceCluster] = cluster
```
better to extract GetLabels() into a local ver

davidohana commented on pkg/mcs/lighthouse.go:
```diff
@@ -0,0 +1,91 @@
+func (*Lighthouse) ListFilter(objmd metav1.ObjectMeta) (*client.ListOptions, error) {
+	if objmd.GetLabels()[lhconst.LabelSourceNamespace] == "" ||
+		objmd.GetLabels()[lhconst.LighthouseLabelSourceName] == "" {
+		return nil, errors.New("missing lighthouse labels")
```
do we want to provide a context of the object with the error?

elevran replied:
will add the object name to the error.

davidohana commented on pkg/mcs/lighthouse.go:
```diff
@@ -0,0 +1,91 @@
+// ListFilter creates a filter that can be used to list only matching ServiceExport
+// or ServiceImport objects. It relies on the presence of the Lighthouse labels.
+func (*Lighthouse) ListFilter(objmd metav1.ObjectMeta) (*client.ListOptions, error) {
+	if objmd.GetLabels()[lhconst.LabelSourceNamespace] == "" ||
```
again, extract GetLabels() into a var

davidohana reviewed (commented):
completed second code review

elevran committed:
move lighthouse naming to lhutils package

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
addressing review comments

- move lighthouse naming convention functions to lhutil package
- consistent use of labels, avoiding use of function call when
  more than one label is accessed/set

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
Address PR review comments

- use `lhutil` package for lighthouse naming convention functions
- pass service name and namespace to reconcile instead of SE object
- move addition of finalizer to Reconcile

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran committed:
added empty lines between object creation and use

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>
